ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ObjectDiffusion/Inbound.hs

```diff
@@ -0,0 +1,448 @@
+        -- Otherwise, we can continue the protocol normally.
+        _continue -> case n of
+          -- We didn't pipeline any requests, so there are no replies in flight
+          -- (nothing to collect)
+          Zero -> do
+            if canRequestMoreObjects st
+              then do
+                -- There are no replies in flight, but we do know some more objects
+                -- we can ask for, so lets ask for them and more objectIds in a
+                -- pipelined way.
+                traceWith tracer $
+                  TraceObjectDiffusionInboundCanRequestMoreObjects (natToInt n)
+                pure $! checkState st & goReqObjectsAndObjectIdsPipelined Zero
+              else do
+                -- There's no replies in flight, and we have no more objects we can
+                -- ask for so the only remaining thing to do is to ask for more
+                -- objectIds. Since this is the only thing to do now, we make this a
+                -- blocking call.
+                traceWith tracer $
+                  TraceObjectDiffusionInboundCannotRequestMoreObjects (natToInt n)
+                pure $! checkState st & goReqObjectIdsBlocking
+
+          -- We have pipelined some requests, so there are some replies in flight.
+          Succ n' ->
+            if canRequestMoreObjects st
+              then do
+                -- We have replies in flight and we should eagerly collect them if
+                -- available, but there are objects to request too so we
+                -- should *not* block waiting for replies.
+                -- So we ask for new objects and objectIds in a pipelined way.
+                traceWith tracer $
+                  TraceObjectDiffusionInboundCanRequestMoreObjects (natToInt n)
+                pure $!
+                  CollectPipelined
+                    (Just (checkState st & goReqObjectsAndObjectIdsPipelined (Succ n')))
+                    (\collected -> checkState st & goCollect n' collected)
+              else do
+                traceWith tracer $
+                  TraceObjectDiffusionInboundCannotRequestMoreObjects (natToInt n)
+                -- In this case we can theoretically only collect replies or request
+                -- new object IDs.
+                --
+                -- But it's important not to pipeline more requests for objectIds now
+                -- because if we did, then immediately after sending the request (but
+                -- having not yet received a response to either this or the other
+                -- pipelined requests), we would directly re-enter this code path,
+                -- resulting us in filling the pipeline with an unbounded number of
+                -- requests.
+                --
+                -- So we instead block until we collect a reply.
+                pure $!
+                  CollectPipelined
+                    Nothing
+                    (\collected -> checkState st & goCollect n' collected)
```

Comment:
NIT: might want to eta-reduce this one as well to match the rest of the callbacks

EDIT: ditto for L264